lisa/tests/scheduler/load_tracking.py

```diff
@@ -958,6 +957,15 @@ def get_trace_cpu_util(self):
         migr_task = sorted(task for task in tasks if task.startswith('migr'))[0]
         migr_task = self.rtapp_task_ids_map[migr_task][0]
 
+        df_sw = self.trace.df_event('sched_switch')
+        df_sw = df_sw[df_sw['next_comm'].str.startswith('swapper')]
+
+        df_state = {}
+
+        for task in tasks:
+            comm = self.rtapp_task_ids_map[task][0].comm
+            df_state[task] = df_sw[(df_sw['prev_comm'] == comm) & (df_sw['prev_state'] == TaskState.TASK_INTERRUPTIBLE)]
```

Comment:
The only difference is that df_task_activation() also contains events when a test task is switched out for another test task, and not only for the idle task. I thought this will mess up the calculation in lisa.pelt.kernel_util_mean() but it looks like this isn't the case. Tested only on one particular run on Juno-r0 so far.

```
-        df_sw = self.trace.df_event('sched_switch')
-        df_sw = df_sw[df_sw['next_comm'].str.startswith('swapper')]
-
         df_state = {}
 
         for task in self.rtapp_task_ids_map.keys():
             comm = self.rtapp_task_ids_map[task][0].comm
-            df_state[task] = df_sw[(df_sw['prev_comm'] == comm) & (df_sw['prev_state'] == TaskState.TASK_INTERRUPTIBLE)]
+            df_state[task] = self.trace.analysis.tasks.df_task_activation(comm)
+            df_state[task] = df_state[task][df_state[task]['active'] == 0]
```

Reply:
> also contains events when a test task is switched out for another test task, and not only for the idle task

It does, there is a `preempted_value` parameter you can set to control what value you will get in this case. By default, it is NaN.

The snippet be turned into this to make it more obvious what is loop-local state and what is the actual content of the df_state is:
```
def filter_state(task_id):
	df = self.trace.analysis.tasks.df_task_activation(task_id)
	return df[df['active'] == 0]

df_state[task] = {
	task: filter_state(task_ids[0])
	for task, task_ids in self.rtapp_task_ids_map.items()
}
```

Also as a general rule, if a TaskID is available, it should always be used instead of a part of it (comm or PID), with `df_filter_task_ids()` if necessary:
https://lisa-linux-integrated-system-analysis.readthedocs.io/en/master/misc_utilities.html#lisa.datautils.df_filter_task_ids

Reply:
I switched to use self.trace.analysis.tasks.df_task_activation in the new version.